feat(CodeActAgent): Support Agent-User Interaction during Task Execution and the Full Integration of CodeActAgent

xingyaoww opened the pull request:
**Dependency**:
- https://github.com/OpenDevin/OpenDevin/pull/1255
- https://github.com/OpenDevin/OpenDevin/pull/1305
- https://github.com/OpenDevin/OpenDevin/pull/1333
- https://github.com/OpenDevin/OpenDevin/pull/1334
- https://github.com/OpenDevin/OpenDevin/pull/1426
- https://github.com/OpenDevin/OpenDevin/pull/1460


For general OpenDevin components:
- Add `IPythonRunCellAction` that allows the agent to use Jupyter plugin introduced in #1255 for task solving.
- `AgentTalkAction` that allows the agent to communicate with the user (either to communicate results OR ask for clarification)
- Agent-User Interaction: Now allow agent's wait for user's response before proceed to task completion.
- Added `__str__` method to some of the actions to get the printed in a readable way
- Add `thought` attribute to most `Action` (if applicable)

For CodeActAgent:
- Allow CodeActAgent to do two things for environmental interactions: (1) use `ipython` AND (2) use `bash`
- Tweak the prompts of CodeActAgent to ensure its generality.
- The ability to use SWE-Agent's tool for task solving (still need more extensive testing).

For integration tests:
- Added integration tests for CodeActAgent.
- Support STDIN mock for "user responses" in integration test

---

The best thing about `CodeActAgent` is that we have existing [data](https://huggingface.co/datasets/xingyaoww/code-act) and open models (e.g., [CodeActAgent](https://huggingface.co/xingyaoww/CodeActAgent-Mistral-7b-v0.1), [CodeQwen](https://huggingface.co/Qwen/CodeQwen1.5-7B-Chat)) trained on that data, which could means we can have a usable local model that can run on laptop.


Below demo is completed with CodeQwen, here's my `config.toml`:
```
LLM_MODEL="ollama/codeqwen"
LLM_API_KEY="ollama"
LLM_BASE_URL="http://localhost:11434"
LLM_EMBEDDING_MODEL="local"
WORKSPACE_BASE="./workspace"
```

![image](https://github.com/OpenDevin/OpenDevin/assets/38853559/480560d0-4957-4d86-a367-7478337d5010)

xingyaoww commented on opendevin/controller/agent_controller.py:
```diff
@@ -150,11 +155,42 @@ def get_task_state(self):
+            user_message_observation = await self._await_user_message_queue.get()
+            self._await_user_message_queue.task_done()
+        return user_message_observation
+
```
@rbren @enyst here's the implemtation of "wait for user info" logic i discussed in https://opendevin.slack.com/archives/C06QKSD9UBA/p1713775908496249

assertion replied:
When the server side waiting for a user input, but user just close the window, how we process with this? What if it's just a  network break for a while, and the browser side come back again seconds later?  @xingyaoww

rbren replied:
Good questions!

We probably need some logic to clean up sessions that have been idle for > x seconds

xingyaoww commented on opendevin/controller/agent_controller.py:
```diff
@@ -190,6 +226,15 @@ async def step(self, i: int):
+            self.add_history(action, user_message_observation)
+            await self._run_callbacks(user_message_observation)
+            return False
+
```
It will wait for user input whenever the agent emits an `AgentTalkAction`

xingyaoww commented on opendevin/server/agent/agent.py:
```diff
@@ -182,14 +182,18 @@ async def start_task(self, start_event):
+            await self.controller.add_user_message(
+                UserMessageObservation(start_event['args']['task'])
+            )
+            return
```
If the agent task is currently running, we directly send an `UserMessageObservation` to the agent instead of killing the task and re-start. Each agent can choose to keep that behavior by implementing separate logic to handle `UserMessageObservation`.

xingyaoww commented on opendevin/controller/agent_controller.py:
```diff
@@ -150,11 +155,42 @@ def get_task_state(self):
+            await self.notify_task_state_changed()
+
+        elif self._task_state == TaskState.RUNNING:
+            self.add_history(NullAction(), message)
```
Will add `NullAction()` with the message, if the agent did not actively ask for user input, but busying executing tasks.

xingyaoww commented:
**UPDATE**
CodeActAgent now can write a calculator and start a flask server in 5 steps! It is also integrated with the ability to discuss with users for clarifing questions, which, IMO is more natural for AI-user interaction.

Here's my `config.toml` to reproduce the following example.
```
LLM_MODEL="gpt-4-turbo"
LLM_API_KEY="sk-TODO"
LLM_EMBEDDING_MODEL=""
WORKSPACE_BASE="./workspace"
SANDBOX_CONTAINER_IMAGE="ghcr.io/opendevin/sandbox:xw-swe-agent-tool-plugins"
```

Command to run: 
`python ./opendevin/main.py -i 10 -c CodeActAgent --task "can you help me start a flask server that build a calculator at 8000?"`

<img width="1452" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/38853559/f5c98ffe-bea7-44a1-8123-ad9b630b45d6">
<img width="1409" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/38853559/582e3610-93ca-4a7b-b966-2af7683b6b04">
<img width="1424" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/38853559/16f9fcf8-e8a2-4f6c-a608-d0d59407583f">

xingyaoww commented:
Finally, I fixed some issues with the front end! This PR should be ready for review!

Here's a video demo for the front end with CodeActAgent (no speed-up).

https://github.com/OpenDevin/OpenDevin/assets/38853559/5563f20a-c1ac-483b-a48e-56968090cb13

There's a bunch of issues for frontend that need to get fixed:

- Sometimes, the model would ask for the user's approval before continuing to execute the code; we should show a button "[Continue]" that sends over the text 'continue' to the agent when "Agent is awaiting user input." It may save some efforts for the user to type the word "[Continue]."

<img width="556" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/38853559/5afc19d4-e260-4d8e-9840-f4d063660a1d">

- IPython code execution is recognized as `undefined` by the front end. We will need to make `Observation` from `IPythonAction` a special class and handle it differently on the front end. I will work on this in a separate PR probably for clarity. Also, the front-end needs to support visualization of `IPythonAction`, potentially by adding an additional tab between `CodeEditor` and `Browser`.

<img width="612" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/38853559/bf712353-2509-40b8-97be-552799cc8e12">

- We need to let the user know things are running! There's a bunch of logs/info on the backend; maybe we can consider displaying some of them to the user so they know the agent is working. Also, some animations focus on showing which panel is changing, which could be helpful for the user. We may have to get token streaming work at some point - otherwise, it takes a ton of time for the user to wait for a `calculator_app.py` to be written.

- We need to get the browser working and monitor the chat between the agent & user and the command line. When it detects URL patterns like `http://127.0.0.1:8000`, a window should pop up a window asking the user whether to open that URL in the built-in browser.

- Around 01:47, we can see the terminal cannot be scrolled down to the bottom. There are probably some CSS issues.

rbren commented on agenthub/codeact_agent/codeact_agent.py:
```diff
@@ -103,19 +81,32 @@ def step(self, state: State) -> Action:
 
         if len(self.messages) == 0:
             assert state.plan.main_goal, 'Expecting instruction to be set'
+            # print(SYSTEM_MESSAGE)
```
Let's remove these debug comments

rbren commented on agenthub/README.md:
```diff
@@ -26,13 +26,15 @@ The `state` contains:
 Here is a list of available Actions, which can be returned by `agent.step()`:
 - [`CmdRunAction`](../opendevin/action/bash.py) - Runs a command inside a sandboxed terminal
 - [`CmdKillAction`](../opendevin/action/bash.py) - Kills a background command
+- [`IPythonRunCellAction`](../opendevin/action/bash.py) - Execute a block of Python code interactively (in Jupyter notebook) and receives `CmdOutputObservation`. Requires setting up `jupyter` [plugin](../opendevin/sandbox/plugins) as a requirement.
```
Can we make this what we called a "pseudo-action" or "auxiliiary action" in your other PR? i.e. keep it internal to CodeAct, and just split it into Write and Run actions under the hood?

xingyaoww replied:
I'm still somewhat debating this -- I guess it ultimately comes down to **whether we want to visualize Jupyter code execution in our frontend**. If so, keeping them might be valuable.

You can see the original CodeAct video for one use case I'm imagining: **plotting figures!** This is already support with the current implementations, we just need to visualize the cell output in our front-end.

These figures are images encoded in `base64` from Jupyter. If we did not do any special handling for Jupyter (i.e., `IPythonRunCellAction` and `IPythonRunCellObservation`), these `base64` images would be interpreted as CmdOutput and get printed to the terminal output, which can be extremely messy.

<video src="https://github.com/xingyaoww/code-act/assets/38853559/62c80ada-62ce-447e-811c-fc801dd4beac"> </video>


If our initial MVP baseline is a ChatGPT-style Python interpreter + Bash command line, I think it could make sense for us to incorporate `IPython Jupyter` as a core action directly: It augments `bash` to complete different sets of tasks (e.g., analysis, visualization).

Side note: If we decide to keep this `IPythonRunCellAction` as the core action, it will probably also be helpful if we pre-install the `jupyter` dependencies in the sandbox image to save the user's time from downloading.

WDYT?

rbren replied:
Let's keep it for now. I'm also good w/ adding jupyter to the sandbox

rbren commented on opendevin/controller/agent_controller.py:
```diff
@@ -159,11 +162,42 @@ def get_task_state(self):
+        if self.state is None:
+            return
+
+        if self._task_state == TaskState.AWAITING_USER_INPUT:
```
Ohh I've been wanting this forever!

li-boxuan replied:
+1

rbren commented on opendevin/server/agent/agent.py:
```diff
@@ -93,7 +93,7 @@ async def dispatch(self, action: str | None, data: dict):
                 await self.create_controller(data)
             case ActionType.START:
-                await self.start_task(data)
+                await self.start_task_or_chat(data)
```
We shouldn't be sending a `START` if it's a chat interruption--I think this should be a separate `case`

rbren replied:
There's currently a status indicator on the frontend, which shows agent status--so it's there somewhere!

rbren reviewed (commented):
Very excited for this!

rbren commented on containers/app/Dockerfile:
```diff
@@ -31,7 +31,7 @@ FROM python:3.12-slim as runtime
 
 ENV RUN_AS_DEVIN=false
-ENV USE_HOST_NETWORK=false
+ENV USE_HOST_NETWORK=true
```
Is this a necessary part of this PR?

I'm going to see if it works for me on latest docker on mac

xingyaoww replied:
revert!

rbren replied:
Hmm I'm still seeing `USE_HOST_NETWORK=true` here

xingyaoww replied:
This is the results of https://github.com/OpenDevin/OpenDevin/pull/1426, which i guess will be the dependency for this one

rbren commented on opendevin/action/bash.py:
```diff
@@ -1,17 +1,23 @@
 @dataclass
 class CmdRunAction(ExecutableAction):
     command: str
+    thought: str = ''
```
Why is there a `thought` on CmdRunAction?

rbren replied:
I've been thinking we should let the LLM override `message` on all the actions it produces, if that helps here

xingyaoww replied:
Because for each bash, CodeActAgent will produce a corresponding thought there.. I think it will be pretty messy if we need to emit two action at the same time (ThoughtAction followed by `CmdRun`), so i just added a `thought` here accompanying CmdRun..

xingyaoww replied:
well, i was thinking even if we want to use `.message` to covey `thought`, we'd still need an additional argument called `thought` or `message` so we can implement the `.message` attribute appropriately

rbren commented on opendevin/config.py:
```diff
@@ -134,6 +135,17 @@ def finalize_config():
         parts = config[ConfigType.WORKSPACE_MOUNT_REWRITE].split(':')
         config[ConfigType.WORKSPACE_MOUNT_PATH] = base.replace(parts[0], parts[1])
 
+    if config.get(ConfigType.WORKSPACE_MOUNT_PATH) is None:
+        config[ConfigType.WORKSPACE_MOUNT_PATH] = os.path.abspath(config[ConfigType.WORKSPACE_BASE])
+
+    USE_HOST_NETWORK = config[ConfigType.USE_HOST_NETWORK].lower() != 'false'
+    if USE_HOST_NETWORK and platform.system() == 'Darwin':
+        logger.warning(
+            'Please upgrade to Docker Desktop 4.29.0 or later to use host network mode on macOS. '
+            'See https://github.com/docker/roadmap/issues/238#issuecomment-2044688144 for more information.'
+        )
+    config[ConfigType.USE_HOST_NETWORK] = USE_HOST_NETWORK
```
I think we've done this differently in other PRs, so maybe this can get removed now?

xingyaoww commented:
It turns out I broke the integration tests by adding a `thought` attribute to most actions. Now, I finally fixed it!

I also added integration tests for CodeActAgent and supported the STDIN mockup for "user responses" during integration tests. (cc @li-boxuan)

CodeActAgent will fail with the `exec` integration task (well, and it is not designed to work with the `exec` box since it is not stateful). @rbren and I were thinking about deprecating `ExecBox` soon, so I removed it from the integration test for now.

This should be ready for review.

li-boxuan commented on agenthub/codeact_agent/codeact_agent.py:
```diff
@@ -93,56 +85,82 @@ def step(self, state: State) -> Action:
+            stop=[
+                '</execute_ipython>',
+                '</execute_bash>',
+                # chatML in case ollama does not stop
```
(nit) what's specific about ollama here? Are you referring to weaker LLMs in general here?

xingyaoww replied:
This was more like an ollama stopword issue that happend while i was testing. it is no long happening, i'm removing it. Thanks for the reminder!

li-boxuan commented on agenthub/codeact_agent/codeact_agent.py:
```diff
@@ -93,56 +85,82 @@ def step(self, state: State) -> Action:
-        command = re.search(r'<execute>(.*)</execute>', action_str, re.DOTALL)
-        if command is not None:
+        bash_command = re.search(r'<execute_bash>(.*)</execute_bash>', action_str, re.DOTALL)
+        python_code = re.search(r'<execute_ipython>(.*)</execute_ipython>', action_str, re.DOTALL)
```
(nit) we don't have to run regex for `python_code` if `bash_command` matches.

xingyaoww replied:
done! fixed with walrus operator!

li-boxuan commented on agenthub/codeact_agent/codeact_agent.py:
```diff
@@ -93,56 +85,82 @@ def step(self, state: State) -> Action:
+            # a code block was found
+            code_group = python_code.group(1).strip()
+            thought = action_str.replace(python_code.group(0), '').strip()
+            # print(f"PYTHON CODE: '{code_group}'")
```
(nit) can be cleaned now?

xingyaoww replied:
done!

li-boxuan commented on agenthub/codeact_agent/codeact_agent.py:
```diff
@@ -93,56 +85,82 @@ def step(self, state: State) -> Action:
-            return AgentEchoAction(
-                content=INVALID_INPUT_MESSAGE
-            )  # warning message to itself
+            # We assume the agent is GOOD enough that when it returns pure NL,
```
Not sure if I understand it. Do you mean "model is GOOD enough..."?

li-boxuan commented on agenthub/codeact_agent/prompt.py:
```diff
@@ -0,0 +1,222 @@
+        SWEAgentCommandsRequirement.documentation.split('\n')
+    )
+)
+# _SWEAGENT_BASH_DOCS content below:
```
Please take this as a grain of salt: my understanding is this paragraph is auto-generated, and you copy & paste here for developers' understanding. There are two reasons again doing this:

1) The content could be found in integration test artifacts.
2) It's hard to keep this comment in sync with its real source.

That being said, this is handy for people to understand, so I don't have a strong opinion.

xingyaoww replied:
Good point! Our project is still under active development; so i guess for now this may benefits devs more in terms of clarity. Once we finalize everything, we can figure out a better way to display this / remove this?

li-boxuan replied:
Sounds fair

li-boxuan commented on agenthub/codeact_agent/prompt.py:
```diff
@@ -0,0 +1,222 @@
+)
+
+SYSTEM_MESSAGE = f"""A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.
+The assistant can interact with an interactive Python (Jupyter Notebook) environment and receive the corresponding output when needed. The code should be enclosed using "<execute_ipython>" tag, for example: <execute_ipython> print("Hello World!") </execute_ipython>.
```
This might not confuse LLM but it confused me 😛 . The example seems to imply it is fine to mess with indentation for `<execute_ipython>` tag. Note the space before and after the `print` statement in the old example.

```suggestion
The assistant can interact with an interactive Python (Jupyter Notebook) environment and receive the corresponding output when needed. The code should be enclosed using "<execute_ipython>" tag, for example: <execute_ipython>print("Hello World!")</execute_ipython>.
```

xingyaoww replied:
Thanks for the suggestion!

I actually changed it to:
```
<execute_ipython>
print("Hello world!")
</execute_ipython>
```

Since, likely, putting `>p` together without space might potentially trigger some weird artifacts/issues in LLM sub-world tokenizers - just to be save :)

li-boxuan replied:
yeah I think that's even better

li-boxuan commented on agenthub/codeact_agent/prompt.py:
```diff
@@ -0,0 +1,222 @@
+"""
+
+_COMMAND_DOCS = (
+    '\nApart from the standard bash commands, you can also use the following special commands in bash:\n'
```
> Apart from the standard bash commands

This seems to contradict the SYSTEM_MESSAGE below, where it is mentioned that bash commands shall be wrapped with `<execute_bash>` tags.

xingyaoww replied:
modified the prompt for clarity

li-boxuan commented on agenthub/codeact_agent/prompt.py:
```diff
@@ -0,0 +1,222 @@
+EXAMPLES = """
+--- START OF EXAMPLE ---
+
+USER: Can you create a list of numbers from 1 to 10, and create a web page to display them at port 5000?
```
It's a bit counter-intuitive for the LLM to choose executing python code to write stuff to a file, while it has other choices like standard bash commands.

Questions:
1. Do you by intention guide the LLM to favor `<execute_ipython>` over `<execute_bash>`, or `open` + `edit` commands?
2. (just curious) If you don't provide this example, and ask the same question, would the assistant use `<execute_ipython>`?

xingyaoww replied:
I think the SWE-Agent primitive is a bit odd for file creation; they would create a new file, open it, and use "edit" to dump in content (3 actions/turns in a role), which I felt is pretty weird in a sense so i randomly adopted python for writing lol.

But other than that, I think as long as they can write to file, it probably doesn't hurt too much - I'm pretty open to (and will) explore if there's a better solution without involving parsing more actions.

I have a pretty strong opinion in making agent's agent space minimalistic in a way: I hope agents can do all of their work in `<execute_bash>` and `<execute_ipython>` without producing action calls in a specific format (e.g., JSON) for parsing (which is a total pain to get it right & can constraint the agent's general capability - that's why i didn't use the `FileWriteAction` for this).

li-boxuan replied:
Fair enough!

li-boxuan reviewed (commented):
This is awesome!

li-boxuan commented:
Btw did you encounter any annoying problems when trying to fix integration tests? Anything you think could be improved?

li-boxuan commented:
Another thought 💭 : 

I presume this agent can do many fancier things than other simpler ones (PlannerAgent? MonologueAgent?). You could probably add a test to demonstrate its larger power. That will also increase the code coverage. Right now, CodeActAgent-specific commands are not tested in the only simple naive integration test we have.

xingyaoww commented:
@li-boxuan, thanks for the review!!

> Btw did you encounter any annoying problems when trying to fix integration tests? Anything you think could be improved?

Nope! I was confused at first until I found the README.md - it was very clear & easy for me to get the integration test fixed.

> I presume this agent can do many fancier things than other simpler ones (PlannerAgent? MonologueAgent?). You could probably add a test to demonstrate its larger power. That will also increase the code coverage. Right now, CodeActAgent-specific commands are not tested in the only simple naive integration test we have. This, of course, doesn't need to be addressed in current PR.


Good idea! I think I will do this in a follow-up PR to ask for more challenging tasks that require multiple interactions with BOTH bash & python.

There will be a lot of TODOs following this:

- Get frontend working well with the new IPython feature https://github.com/OpenDevin/OpenDevin/pull/1363
- Add more integration tests for more complex behaviors (multi-turn interactions with both bash & python)
- Get SWE-Bench ready for testing (prompts are also subject to change)

rbren commented on .github/workflows/run-integration-tests.yml:
```diff
@@ -43,11 +43,6 @@ jobs:
             agent: "MonologueAgent"
             embedding-model: "local"
             sandbox: "exec"
-          - name: CodeActAgent-py311-exec
-            python-version: "3.11"
-            agent: "CodeActAgent"
-            embedding-model: "none"
-            sandbox: "exec"
```
Why are we removing this test case? I'd like to keep intercompatibility of agents and sandbox envs--is there a particular reason CodeAct can't work with exec?

xingyaoww replied:
As discussed above: 

> CodeActAgent will fail with the exec integration task (well, and it is not designed to work with the exec box since it is not stateful). @rbren and I were thinking about deprecating ExecBox soon, so I removed it from the integration test for now.

CodeActAgent is not designed for execbox, even if we barely get it work there, it won't work well. e.g., CodeActAgent needs `source ~/.bashrc` to load plugins and tools (e.g., swe-agent primitives) to function properly. But **ExecBox is not stateful**, the commands/tools you registered in a previous execution is not available in the next `.execute` request, essentially make the whole thing not working.

I don't see a point of spending too much efforts in making CodeActAgent compatible with ExecBox, especially we are already considering deprecating it.

xingyaoww replied:
I'll double-check to confirm that this PR will work well with DooD + SSHBox, and hopefully it will mitigate the issue of not being able to use ExecBox.

li-boxuan replied:
You probably need to add a note to user doc that this agent doesn't work with execbox. Or even throw an exception.

li-boxuan replied:
According to https://github.com/OpenDevin/OpenDevin/pull/1149 @foragerr, seems we cannot use sshbox for dogfood purpose?

xingyaoww replied:
I'll do more extensive testing on DooD + SSHBox, and i think for dogfood in Github, we could also use `LocalBox` for this, which i do plan to test as well.

li-boxuan replied:
That's a good point. LocalBox should be enough for dogfood CI to run.

xingyaoww commented:
Finally, with some modifications (https://github.com/OpenDevin/OpenDevin/pull/1426), I am finally able to get CodeActAgent working in the DooD sandbox with `RUN_AS_DEVIN` and `network=host` without issues :)

Steps to reproduce:
1. switch to this PR
2. build the `app` container: `./containers/build.sh sandbox` (you may need to comment a few lines in `build.sh` to make it work on your local machine
3. run the following, switch to CodeActAgent on the web interface, and it should work:
```
docker run \
    -e LLM_API_KEY \
    -e WORKSPACE_MOUNT_PATH=$WORKSPACE_BASE \
    -v $WORKSPACE_BASE:/opt/workspace_base \
    -e SANDBOX_USER_ID=$(id -u) \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -p 3000:3000 \
    --add-host host.docker.internal=host-gateway \
    ghcr.io/opendevin/opendevin:latest
```

rbren commented on containers/app/Dockerfile:
```diff
@@ -32,21 +32,32 @@ FROM python:3.12-slim as runtime
-    && apt-get install -y curl ssh
+    && apt-get install -y curl ssh sudo
+
+RUN useradd -m -u $SANDBOX_USER_ID -s /bin/bash opendevin && \
+    usermod -aG sudo opendevin && \
+    echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
+RUN chown -R opendevin:opendevin /app
+USER opendevin
```
Can you explain why the sandbox user is in the app container? That doesn't seem right to me....

xingyaoww replied:
This is something i did in https://github.com/OpenDevin/OpenDevin/pull/1426

maybe we can merge this first. The reason is that, when mounted dir in DooD setup, if the user is `root` but not `SANDBOX_USER_ID`, there's a chance that the `app` server will encounter permission denied error in writing to `/opt/workspace_base`.

rbren replied:
👍 thanks for the explanation! It does make sense. I hope we figure out a way to do this better in the future but this definitely makes sense for now

rbren commented on opendevin/config.py:
```diff
@@ -49,7 +49,7 @@
     ConfigType.E2B_API_KEY: '',
     ConfigType.SANDBOX_TYPE: 'ssh',  # Can be 'ssh', 'exec', or 'e2b'
-    ConfigType.USE_HOST_NETWORK: 'false',
+    ConfigType.USE_HOST_NETWORK: 'true',
```
```suggestion
    ConfigType.USE_HOST_NETWORK: 'false',
```

xingyaoww replied:
fixed!

rbren commented on opendevin/config.py:
```diff
@@ -167,10 +167,6 @@ def finalize_config():
         )
     config[ConfigType.USE_HOST_NETWORK] = USE_HOST_NETWORK
 
-    if config.get(ConfigType.WORKSPACE_MOUNT_PATH) is None:
-        config[ConfigType.WORKSPACE_MOUNT_PATH] = config.get(ConfigType.WORKSPACE_BASE)
-
```
Any reason for deleting this? Seems to help some users

rbren commented:
@xingyaoww I implemented the user_message action here: https://github.com/OpenDevin/OpenDevin/commits/rb/codeact/

But I can't find a way to contribute it to your PR...

rbren commented on agenthub/codeact_agent/codeact_agent.py:
```diff
@@ -1,54 +1,37 @@
+    AgentTalkAction,
     CmdRunAction,
+    IPythonRunCellAction,
+    NullAction,
```
```suggestion
    NullAction,
    
```

rbren commented:
@xingyaoww once my PR is merged into yours, this LGTM!

xingyaoww commented:
@rbren merged! this should be ready to go!

rbren commented:
🎉 glad to have this in!

xingyaoww commented:
@rbren Thanks!!!

Next steps:
- Frontend!
  - https://github.com/OpenDevin/OpenDevin/pull/1363/
  - Switch to Jupyter Tab when a python execution request comes in (once https://github.com/OpenDevin/OpenDevin/pull/1363/ is merged)
  - https://github.com/OpenDevin/OpenDevin/issues/1487
- Evaluation: get SWE-Bench score for CodeActAgent!
